packages/ai/core/util/messages-store.ts

```diff
@@ -0,0 +1,348 @@
+import { generateId as generateIdFunction, parsePartialJson } from '.';
+import type {
+  ReasoningUIPart,
+  TextUIPart,
+  ToolInvocation,
+  ToolInvocationUIPart,
+  UIMessage,
+} from '../types';
+import { throttle } from './throttle';
+
+type SubscriberCallback = () => void;
+
+/**
+ * Internal class for managing UIMessages
+ */
+export class MessagesStore {
+  chatId: string;
+  private messages: UIMessage[] = [];
+  private subscribers: Set<SubscriberCallback>;
+  private notify: () => void;
+
+  // Temporary state for in-progress chat response:
+  private partialToolCalls: Record<
+    string,
+    { text: string; step: number; index: number; toolName: string }
+  > = {};
+  private tempParts: {
+    text?: TextUIPart;
+    reasoning?: ReasoningUIPart;
+    reasoningTextDetail?: { type: 'text'; text: string; signature?: string };
+  } = {};
+
+  constructor({
+    chatId = generateIdFunction(),
+    initialMessages,
+    throttleMs,
+  }: {
+    initialMessages: UIMessage[];
+    throttleMs?: number;
+    chatId?: string;
+  }) {
+    this.messages = initialMessages;
+    this.chatId = chatId;
+    this.subscribers = new Set();
+    this.notify = throttleMs
+      ? throttle(() => this.emit(), throttleMs)
+      : () => this.emit();
```

Comment:
how are `emit`, `subscribers`, and `notify` related?

Reply:
subscribers holds all the callbacks we call when we want to notify of changes! notify is a wrapper around the internal emit method that is throttled if throttleMs is passed in!

Reply:
maybe smol doc comments on `notify` and `emit` would be helpful